hivemind/client/averaging/__init__.py

```diff
@@ -279,6 +289,134 @@ async def rpc_aggregate_part(self, stream: AsyncIterator[averaging_pb2.Averaging
+        return await future
+
+    def _background_thread_fetch_current_state_if_asked(self):
+        """ Executed in the host process as a background thread. """
+        while True:
+            trigger, future = self.pipe.recv()
+            assert trigger == '_TRIGGER_GET_CURRENT_STATE'
+            try:
+                state_metadata, state_tensors = self.get_current_state()
+                # note: serialize here to avoid initializing cuda in the guest process
+                state_metadata = pickle.dumps(state_metadata)
+                state_tensors = tuple(tensor.cpu().detach().requires_grad_(tensor.requires_grad)
+                                      for tensor in state_tensors)
+                future.set_result((state_metadata, state_tensors))
+            except BaseException as e:
+                future.set_exception(e)
+                logger.warning(e)
+                continue
+
+    def load_state_from_peers(self, wait=True) -> Optional[Any]:
+        """ Try to download the latest optimizer state one of the existing peer """
+        future, _future = MPFuture.make_pair()
+        self.pipe.send(('_load_state_from_peers', [], dict(future=_future)))
+        return future.result() if wait else future
+
+    async def _load_state_from_peers(self, future: MPFuture):
+        key_manager = self._matchmaking.group_key_manager
+        peers_info, _ = self.dht.get(f"{key_manager.prefix}.all_averagers", latest=True) or ({}, None)
+        metadata = None
+        if not isinstance(peers_info, dict):
+            logger.info("Averager could not load state from peers: peer dict is corrupted.")
+            future.set_result(None)
+            return
+
+        def get_priority(peer):
+            info = peers_info[peer]
+            if not isinstance(info, ValueWithExpiration) or not isinstance(info.value, (float, int)):
+                logger.warning(f"Skipping averager {peer} - bad peer info {info.value} (expected a number).")
+                return -float('inf')
+            return float(info.value)  # prefer peers with latest update time
+
+        for peer in sorted(peers_info.keys(), key=get_priority, reverse=True):
+            if peer != self.endpoint:
+                logger.info(f"Downloading parameters from peer {peer}")
+                stream = None
+                try:
+                    leader_stub = ChannelCache.get_stub(peer, averaging_pb2_grpc.DecentralizedAveragingStub, aio=True)
+                    stream = leader_stub.rpc_download_state(averaging_pb2.DownloadRequest())
+                    current_tensor_parts, tensors = [], []
+                    async for message in stream:
+                        if message.metadata:
+                            metadata = pickle.loads(message.metadata)
+                        if message.tensor_part.dtype and current_tensor_parts:
+                            tensors.append(deserialize_torch_tensor(combine_from_streaming(current_tensor_parts)))
```

Comment:
Needs a comment that `message.tensor_part.dtype` indicates the start of a new tensor